### DISCUSSION TITLE
[Feature request]：关于用户界面的建议

### DISCUSSION THREAD

**Discussion by LuorixDev** [pk5ls20/LiteLoaderQQNT-NekoImage#13] [Ideas]
### 以下是我的一些建议，假如您愿意实现可以尝试一下（）

---

~~QQNT的主要受众可能还是以国内用户为主，是否能支持**汉化**。~~ 已实装

---

是否能够支持**无感下拉**，当进度条不足200xp时候自动发送请求，拉取的尽头添加“终末”提示

![image](https://github.com/pk5ls20/LiteLoaderQQNT-NekoImageGallerySearch/assets/157946924/53b1ac7c-9867-464f-99d7-dbb0073b0643)

---

为表情添加**用户权重**，当该表情为用户选择表情时，给这个图片记录+1的权重数字，当二次搜索时，搜索顺序为“similar+权重”的降序排列，直接保存在Ntqq liteloader data目录下就完全够了，上游应该没有提供类似功能。
`附上条：右键图片允许手动修改权重值or固定权重值、允许用户设置是否在搜索界面每张图片左下角以灰色小字显示similar和权重，允许用户修改每次权重+n，允许用户修改权重的占比值‘similar+权重*k’`

---

~~允许右键删除图片，用户界面可以更方便的看出重复图片~~ 已实装

**Comment by pk5ls20** [comment]
> ### 以下是我的一些建议，假如您愿意实现可以尝试一下（）
> QQNT的主要受众可能还是以国内用户为主，是否能支持**汉化**。
>

https://github.com/pk5ls20/LiteLoaderQQNT-NekoImageGallerySearch/releases/tag/v0.1.0


> 是否能够支持**无感下拉**，当进度条不足200xp时候自动发送请求，拉取的尽头添加“终末”提示
> 
> ![image](https://private-user-images.githubusercontent.com/157946924/344393016-53b1ac7c-9867-464f-99d7-dbb0073b0643.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTk2NjkwOTgsIm5iZiI6MTcxOTY2ODc5OCwicGF0aCI6Ii8xNTc5NDY5MjQvMzQ0MzkzMDE2LTUzYjFhYzdjLTk4NjctNDY0Zi05OWQ3LWRiYjAwNzNiMDY0My5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjI5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYyOVQxMzQ2MzhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jYmUxMjE5MTk4MDZjOTQ3ZWMyNjIxY2M5Zjk4MDQzZTU3NzY0OGMzMmJhZTVlNGUzMWM4OWY3MzA0ZWMzOWJiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.410Asnz41DEUBHnaM1Od7Q9NhV9oqNRHk61PgNE4flA)
> 

也许会做（？

> 为表情添加**用户权重**，当该表情为用户选择表情时，给这个图片记录+1的权重数字，当二次搜索时，搜索顺序为“similar+权重”的降序排列，直接保存在Ntqq liteloader data目录下就完全够了，上游应该没有提供类似功能。 `附上条：右键图片允许手动修改权重值or固定权重值、允许用户设置是否在搜索界面每张图片左下角以灰色小字显示similar和权重，允许用户修改每次权重+n，允许用户修改权重的占比值‘similar+权重*k’`
> 

这个我再考虑下🤔

> 允许右键删除图片，用户界面可以更方便的看出重复图片

https://github.com/pk5ls20/LiteLoaderQQNT-NekoImageGallerySearch/releases/tag/v0.1.0

**Comment by LuorixDev** [comment]
谢谢大佬回复！😚

**Comment by LuorixDev** [comment]
## 一个创意：
能不能检测到两张图片相似度值相差小于2，就以任意一张作为搜索，如果再次搜索第一项相似度达到99%以上，则在用户界面用框框圈出，并自动删除一张
![image](https://github.com/pk5ls20/LiteLoaderQQNT-NekoImageGallerySearch/assets/157946924/765138b9-57cb-40ed-89ae-d31b67b53433)
### 1.发现相似图片【后台自动计算】
![image](https://github.com/pk5ls20/LiteLoaderQQNT-NekoImageGallerySearch/assets/157946924/9b0be577-dc4c-42c6-a675-94a4448aa5f6)
![image](https://github.com/pk5ls20/LiteLoaderQQNT-NekoImageGallerySearch/assets/157946924/4599110c-8a11-4dca-8458-a46ba4dbff87)

### 2.以任意图片搜索【后台搜索不影响用户界面】
![image](https://github.com/pk5ls20/LiteLoaderQQNT-NekoImageGallerySearch/assets/157946924/0c71407d-037a-42ca-bd89-83c5dae33933)

发现最相近图片相似度过高

### 3.在用户界面标出，模拟图：
![image](https://github.com/pk5ls20/LiteLoaderQQNT-NekoImageGallerySearch/assets/157946924/4cf405ea-3fe7-4e3f-bc09-3109d5c5da39)

**Reply by pk5ls20** [reply]
图片重复大致可以分为两种情况：
1. 两张图片**完全一致**下的去重（即两张图片的uuid相同，本项目及其上游项目的uuid的计算方式参见https://github.com/hv0905/NekoImageGallery/blob/0f242b1ad91b08299082f3eddc5db5092d3298bb/app/util/generate_uuid.py ）
在[最新版本中的NekoImageGallery](https://github.com/hv0905/NekoImageGallery/commit/0f242b1ad91b08299082f3eddc5db5092d3298bb)中，上传图片时如果两张图片**完全一致** ，则会返回409 (The uploaded point is already contained in the database)，从而保证不会出现这种情况。

2. 两张（或多张）图片相似度过高下的去重
这种情况下两张（或多张）图片从视觉效果上来看，**有可能完全一致，亦有可能有细微的视觉上的差别**。
理论上来讲，可以使用 感知哈希相似度 / 图片CLIP向量相似度 结合 KNN / ANN 来进行去重，即通过设定一个阈值对大于一定相似度的一对（或一组）图片进行去重。
目前，在上传过程中进行本操作的可行性有待考虑，但是对NekoImageGallery的已有图片进行批量去重目前是完全可行的。
稍晚点我写一个对NekoImageGallery的已有图片进行批量去重的demo

参考：
https://edgeneko.aiursoft.cn/2023/08/06/ImageHashUsage

**Comment by LuorixDev** [comment]
话说现在版本自己的转发消息无法批量上传？

**Reply by pk5ls20** [reply]
是这样的。
因为目前检测及读取合并转发内容基于HookVue，这些无法批量上传的合并转发消息不是正常的`multiForwardMsgElement`，而是`arkElement`
~~后续版本可能会支持这类消息的上传~~
https://github.com/pk5ls20/LiteLoaderQQNT-NekoImageGallerySearch/commit/7cceb62fc92c0fdcd8fc00faf25c318cb1f5bd13 已实装

**Comment by LuorixDev** [comment]
我看现在的版本没办法多选上传，于是想出可以先多选表情包合并转发给自己，让后发现自己的消息无法批量上传

**Reply by pk5ls20** [reply]
~~诶？也许以后有时间可以做一下多选上传（画饼ing~~
已初步实现于: https://github.com/pk5ls20/LiteLoaderQQNT-NekoImageGallerySearch/commit/4f88749148a091bcab4a2cc796de7dcbf6390143
![d5d01a30c0dd7b29811f95102c23899a](https://github.com/user-attachments/assets/2b83aec0-c330-4663-b827-825f86b20ded)

**Comment by LuorixDev** [comment]
![image](https://github.com/user-attachments/assets/c8440ae4-371c-435b-bf16-ca09f214b3ec)

**Reply by pk5ls20** [reply]
https://github.com/pk5ls20/LiteLoaderQQNT-NekoImageGallerySearch/commit/7cceb62fc92c0fdcd8fc00faf25c318cb1f5bd13
